### PR TITLE
Feature/hex string utils

### PR DESCRIPTION
**Pull request by james-toussaint**

### CODE REVIEW

File: src/main/java/com/iexec/common/utils/BytesUtils.java [hunk 1]
```diff
@@ -44,9 +45,32 @@ public static String hexStringToAscii(String hexString) {
...
+     * Hexadecimal string input must be lowercase and must contain 0x prefix.
+     *
+     * @param hexString        hexadecimal input
+     * @param expectedByteSize expected byte size
+     * @return true if
+     */
+    public static boolean isHexStringWithProperByteSize(String hexString,
+                                                        int expectedByteSize) {
+        return expectedByteSize > 0
+                && !StringUtils.isEmpty(hexString)
```
Comment (zguesmi) [inline src/main/java/com/iexec/common/utils/BytesUtils.java]:
```suggestion
                && StringUtils.isNotEmpty(hexString)
```

File: src/main/java/com/iexec/common/utils/BytesUtils.java [hunk 2]
```diff
@@ -44,9 +45,32 @@ public static String hexStringToAscii(String hexString) {
...
+                && Numeric.containsHexPrefix(hexString)
+                && isHexaString(hexString)
+                && BytesUtils.stringToBytes(hexString).length == expectedByteSize;
```
Comment (zguesmi) [inline src/main/java/com/iexec/common/utils/BytesUtils.java]: I find method `BytesUtils.stringToBytes(hexString)` confusing. The name gives the impression that it has the same behavior as `String.getBytes()` but it is not. I think it's better to either directly use `Numeric.hexStringToByteArray(hexString)` or rename it to something like `hexStringToBytes()`. I tend to prefer the first option as the method itself seems to not be very useful.

File: src/test/java/com/iexec/common/utils/BytesUtilsTests.java [hunk 3]
```diff
@@ -61,6 +61,50 @@ public void shoudNotBeHexaStringSinceEmpty() {
...
+        assertFalse(isHexStringWithProperByteSize("", 6));
+    }
+
+    @Test
+    void shouldNotBeHexStringWithProperByteSizeSinceNoPrefix() {
+        assertFalse(isHexStringWithProperByteSize("abc123", 6));
+    }
+
+    @Test
+    void shouldNotBeHexStringWithProperByteSizeSinceNotLowerCase() {
```
Comment (zguesmi) [inline src/test/java/com/iexec/common/utils/BytesUtilsTests.java]: Why is that? In this case shouldn't the method be named `isLowercaseHexStringWithProperByteSize()`

Comment (james-toussaint) [reply src/test/java/com/iexec/common/utils/BytesUtilsTests.java]: This is what I was updating, let me send you a next version

File: src/main/java/com/iexec/common/utils/BytesUtils.java [hunk 4]
```diff
@@ -44,9 +46,42 @@ public static String hexStringToAscii(String hexString) {
     }
 
     public static boolean isHexaString(String hexaString) {
-        return Numeric.cleanHexPrefix(hexaString).matches("\\p{XDigit}+"); // \\p{XDigit} matches any hexadecimal character
+        return !StringUtils.isEmpty(hexaString)
```
Comment (mcornaton) [inline src/main/java/com/iexec/common/utils/BytesUtils.java]:
```suggestion
        return StringUtils.isNotEmpty(hexaString)
```

**Review by mcornaton — APPROVED** [review]

File: src/main/java/com/iexec/common/utils/BytesUtils.java [hunk 5]
```diff
@@ -44,9 +46,43 @@ public static String hexStringToAscii(String hexString) {
...
+                new byte[expectedByteSize]);
+    }
+
+    public static boolean isBytes32(String hexString) {
+        return isHexStringWithProperBytesSize(hexString, 32);
+    }
+
+    public static boolean isNonEmptyBytes32(String hexString) {
+        return isHexStringWithNonEmptyProperBytesSize(hexString, 32);
+    }
```
Comment (jbern0rd) [inline src/main/java/com/iexec/common/utils/BytesUtils.java:98]: Should we introduce a `private static final int PROPER_BYTES_SIZE = 32` numeric constant ?

File: src/main/java/com/iexec/common/utils/BytesUtils.java [hunk 6]
```diff
@@ -44,9 +46,43 @@ public static String hexStringToAscii(String hexString) {
...
+     *
+     * @param hexString        hexadecimal input
+     * @param expectedByteSize expected byte size
+     * @return true if
+     */
+    public static boolean isHexStringWithProperBytesSize(String hexString,
+                                                        int expectedByteSize) {
+        return expectedByteSize > 0
+                && StringUtils.isNotEmpty(hexString)
+                && Numeric.containsHexPrefix(hexString)
```
Comment (jbern0rd) [inline src/main/java/com/iexec/common/utils/BytesUtils.java]: `StringUtils.isNotEmpty` will be called another time in `isHexaString`.

Comment (jbern0rd) [reply src/main/java/com/iexec/common/utils/BytesUtils.java]: Why not simply `hexString.matches("^0x\\p{XDigit}+$")` ?
If there is a match:
* it is not empty
* it has the prefix
* we don't need to remove the prefix to check if it is correct

Or am I missing something entirely ?

Comment (jbern0rd) [reply src/main/java/com/iexec/common/utils/BytesUtils.java]: For improved performance, the best implementation should look like:

`private static final Pattern hexPattern = Pattern.compile(<regex>);`

then

`hexPattern.matcher(hexString).matches()`

This avoids to compile the regex pattern every time `isHexaString` is called

File: src/main/java/com/iexec/common/utils/BytesUtils.java [hunk 7]
```diff
@@ -44,9 +46,43 @@ public static String hexStringToAscii(String hexString) {
...
+                && isHexaString(hexString)
+                && Numeric.hexStringToByteArray(hexString).length == new byte[expectedByteSize].length;
```
Comment (zguesmi) [inline src/main/java/com/iexec/common/utils/BytesUtils.java]: Why not?
```suggestion
                && Numeric.hexStringToByteArray(hexString).length == expectedByteSize;
```

**Review by zguesmi — APPROVED** [review]

**Comment by zguesmi** [comment]
Good :+1:

**Review by mcornaton — APPROVED** [review]
LGTM :)

File: src/main/java/com/iexec/common/utils/BytesUtils.java (hunk 7 above)
Comment (james-toussaint) [reply src/main/java/com/iexec/common/utils/BytesUtils.java]: 8 bits = 1 byte -> 2 char slots (0xff fits in 1 byte)?

Comment (mcornaton) [reply src/main/java/com/iexec/common/utils/BytesUtils.java]: Not sure about what you mean, but `new byte[expectedByteSize].length` is strictly equal to `expectedByteSize` as `new type[n]` creates an array of length `n` and `array.length`retrieves the length of the array...

Comment (james-toussaint) [reply src/main/java/com/iexec/common/utils/BytesUtils.java]: "new byte[expectedByteSize].length is strictly equal to expectedByteSize"
totally agree, let me update that :+1:

File: src/main/java/com/iexec/common/utils/BytesUtils.java (hunk 2 above)
Comment (james-toussaint) [reply src/main/java/com/iexec/common/utils/BytesUtils.java]: I agree, updated thanks

File: src/main/java/com/iexec/common/utils/BytesUtils.java (hunk 5 above)
Comment (james-toussaint) [reply src/main/java/com/iexec/common/utils/BytesUtils.java:98]: Updated thanks

File: src/main/java/com/iexec/common/utils/BytesUtils.java (hunk 6 above)
Comment (james-toussaint) [reply src/main/java/com/iexec/common/utils/BytesUtils.java]: I agree. Updated, how about that https://github.com/iExecBlockchainComputing/iexec-common/pull/279/commits/615c56c352054102b665c6a8575900310edf44ab?

Comment (jbern0rd) [reply src/main/java/com/iexec/common/utils/BytesUtils.java]: :+1: 
Looks great to me, thx !

**Pull request merged**